### DISCUSSION TITLE
Push fails when pushed with `mirror` option and repo contains 1000s of refs in 'refs/replace'

### DISCUSSION THREAD

**Discussion by thejewdude** [community/community#21651] [New to GitHub]
<p>My team and I are about to move all our repos to Github but I’m having trouble pushing our main repo. As part of the migration we shrunk our repo size by rewriting the git history to remove large files that we no longer need. The tool we used, git-filter-repo, created a 1000s of replace refs under <code>refs/replace</code> and when I try pushing the repo with <code>git push origin --mirror</code> I get a <code>remote: Internal Server Error</code>. I attached an image that shows the error.</p>
<p><img src="https://user-images.githubusercontent.com/17015713/181066011-8a1449ab-1c56-4cda-a05b-c94a566e2870.png" alt="git push error" data-base62-sha1="xi3zuC5jR38QfCXqTfYkGdxBHn0" width="608" height="185"></p>
<p>Any help would be appreciated, and it would be great if I can tell what the internal server error is</p>

**Comment by canuckjacq** [comment]
<p>:wave: Welcome!</p>
<p>500 is a hard one to troubleshoot! Have you resolved this yet?</p>

**Comment by thejewdude** [comment, marked as answer]
<p>Sort of.<br>
I reached out to support by email and one of your colleagues indicated that its a known limitation that GitHub will timeout when trying to push too many refs at once. Their suggestion was to push the refs in smaller chunks. I accomplished this with a wee bit of commandline hackery</p>
<pre><code class="lang-auto">git for-each-ref --format="%(refname)" --sort='authordate' refs/replace | xargs git push origin
</code></pre>

**Comment by shahbazmastan** [comment]
<p>Hi <a class="mention" href="/thejewdude">@thejewdude</a></p>
<p>I am facing same error while migrating code from gerrit to github, can you please elaborate what have you done.</p>
<p>TIA</p>

**Comment by thejewdude** [comment]
<p>Basically the command line I posted above pushes the refs in smaller chunks. I’m using the Linux program xargs to automatically chunk pushing the refs (I think it was about 500 at a time) In my case I had &gt; 10,000 replace refs, that’s why my command line mentions refs/replace.</p>
<p>Let me know if something still isn’t clear!</p>

**Comment by shahbazmastan** [comment]
<p>Thanks, it worked for some repos but unfortunately for some its failing<br>
using this command for push<br>
<code>git for-each-ref --format="%(refname)" --sort='authordate' | xargs git push origin</code><br>
before executing the above command i am executing this<br>
<code>git config --unset remote.origin.mirror</code></p>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/3X/7/2/72927c31e54c301c47e7627856bd70437520afe3.png" data-download-href="/uploads/short-url/glyh1w6iA4AjuGUpHjrLXbWYReb.png?dl=1" title="error" rel="noopener nofollow ugc"><img src="https://user-images.githubusercontent.com/81965864/181066017-35222dca-ba94-43c7-b6bb-cefdff19c802.png" alt="error" data-base62-sha1="glyh1w6iA4AjuGUpHjrLXbWYReb" width="690" height="140" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/3X/7/2/72927c31e54c301c47e7627856bd70437520afe3_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">error</span><span class="informations">820×167 9.07 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>

**Comment by thejewdude** [comment]
<p>What does unsetting the mirror do?<br>
Is it possible that you’re pushing refs that already exist on the remote? If you’re pushing to the same repo that tried to push to earlier (and failed), it be causing problems. You might want to try my command line with a fresh repo.<br>
You should also try changing the amount of args that <code>xargs</code> will run at a time. For me it worked with &lt; 500 but YMMV</p>

**Comment by shahbazmastan** [comment]
<p>I am unsetting the mirror because i faced this<br>
<code>error: --mirror can't be combined with refspecs</code>, when i googled some one suggested to unset mirror.<br>
anyhow i pushed the tags and branches separately.<br>
Thanks for response</p>

**Comment by bhalbright** [comment]
FWIW, I added -P 3 to the command from @thejewdude to run some of the pushes in parallel, got my repo to push in about 30 minutes vs 2 hours before. And thanks so much @thejewdude, your script saved me hours of frustration :)

`git for-each-ref --format="%(refname)" --sort='authordate' refs/replace | xargs -P 3 git push origin`
